File: BFS/python/bfs_seq.py
class Graph:
    def __init__(self) -> None:
        self.vertices: dict[int, list[int]] = {}
    
    def add_edge(self, from_vertex: int, to_vertex: int) -> None:
        if from_vertex not in self.vertices:
            self.vertices[from_vertex] = []
        if to_vertex not in self.vertices:
            self.vertices[to_vertex] = []
        self.vertices[from_vertex].append(to_vertex)
        self.vertices[to_vertex].append(from_vertex)


def bfs(graph: Graph, start_vertex: int) -> list[int]:
    if start_vertex not in graph.vertices:
        return []
    
    visited = set()
    result = []
    queue = [start_vertex]
    while queue:
        current = queue.pop(0)
        if current not in visited:
            visited.add(current)
            result.append(current)
            for neighbor in graph.vertices.get(current, []):
                if neighbor not in visited:
                    queue.append(neighbor)

    return result

File: BFS/cpp/bfs_seq.cpp
#include "bfs_seq.hpp"
#include "graph.h"
#include "queue"
#include <iostream>

std::vector<int> bfs(Graph& g, int start_vertex) {
    std::vector<int> result = {};
    
    if (g.vertices.count(start_vertex) == 0){
        return result;
    }

    std::unordered_map<int, bool> visited;
    std::queue<int> q;

    visited[start_vertex] = 1;
    q.push(start_vertex);

    while (!q.empty()) {
        int current = q.front();
        q.pop();
        result.push_back(current);  
        
        for (int neighbor : g.vertices.at(current)) {
            if (!visited[neighbor]) {
                visited[neighbor] = 1;
                q.push(neighbor);
            }
        }
    }

    return result;
}


